main/src/ca/uwaterloo/flix/language/phase/Regions.scala

```diff
@@ -421,16 +422,84 @@ object Regions {
     */
   private def checkType(tpe: Type, loc: SourceLocation)(implicit scope: List[Type.Var], flix: Flix): Validation[Unit, CompilationMessage] = {
     // Compute the region variables that escape.
-    val escapes = regionVarsOf(tpe) -- scope
+    val minned = TypeMinimization.minimizeType(tpe)
+    val regs = regionVarsOf(minned)
+    for (reg <- regs -- scope) {
+      if (essentialTo(reg, minned)) {
+        return TypeError.RegionVarEscapes(reg, minned, loc).toFailure
+      }
+    }
+    ().toSuccess
+  }
+
+  /**
+    * Returns true iff the type variable `tvar` is essential to the type `tpe`.
+    *
+    * A type variable is essential if its ascription has a bearing on the resulting value.
+    * For example, in the type `a and (not a)`, `a` is not essential since the result is always `false`.
+    */
+  def essentialTo(tvar: Type.Var, tpe: Type)(implicit flix: Flix): Boolean = {
+    if (!tpe.typeVars.contains(tvar)) {
+      // Case 1: The type variable is not present in the type. It cannot be essential.
+      false
+    } else {
+      // Case 2: The type variable is present in the type. Check if it is essential to any of the booleans.
+      boolTypesOf(tpe).exists(essentialToBool(tvar, _))
+    }
+  }
 
-    // Return an error if a region variable escapes.
-    if (escapes.nonEmpty) {
-      val rvar = escapes.head
-      return TypeError.RegionVarEscapes(rvar, tpe, loc).toFailure
+  /**
+    * Returns true iff the type variable `tvar` is essential to the boolean formula `tpe`.
+    * Assumes that `tvar` is present in the type.
+    */
+  def essentialToBool(tvar: Type.Var, tpe: Type)(implicit flix: Flix): Boolean = {
+    val t0 = tpe.map {
```

Comment:
This was not clear to me, maybe we can write a small function substituteVar to make it clear.